### PR TITLE
Added Decrypt func to ocr2 OffchainKeyring

### PR DESCRIPTION
**Pull request by KuphJr**

### CODE REVIEW

File: core/services/keystore/keys/ocr2key/offchain_keyring.go [hunk 1]
```diff
@@ -47,6 +50,25 @@ func newOffchainKeyring(encryptionMaterial, signingMaterial io.Reader) (*Offchai
 	return ok, nil
 }
 
+// Decrypt decrypts a message that was encrypted using the public key
+func (ok *OffchainKeyring) Decrypt(ciphertext []byte) (plaintext []byte, err error) {
+	if len(ciphertext) < box.Overhead {
+		return nil, errors.New("ciphertext too short")
+	}
+
+	var nonce [24]byte
```
Comment (pszal) [inline core/services/keystore/keys/ocr2key/offchain_keyring.go:59]: unused?

File: core/services/keystore/keys/ocr2key/offchain_keyring.go [hunk 2]
```diff
@@ -47,6 +50,25 @@ func newOffchainKeyring(encryptionMaterial, signingMaterial io.Reader) (*Offchai
...
+	copy(nonce[:], ciphertext[:box.Overhead])
+
+	publicKey := [curve25519.PointSize]byte(ok.ConfigEncryptionPublicKey())
+
+	decrypted, success := box.OpenAnonymous(nil, ciphertext, &publicKey, &ok.encryptionKey)
```
Comment (pszal) [inline core/services/keystore/keys/ocr2key/offchain_keyring.go:64]: I though we want `Open()` and not `OpenAnonymous()`, isn't that the case? (Then `nonce` would be used.)

File: core/services/keystore/keys/ocr2key/offchain_keyring.go [hunk 3]
```diff
@@ -47,6 +50,25 @@ func newOffchainKeyring(encryptionMaterial, signingMaterial io.Reader) (*Offchai
 	return ok, nil
 }
 
+// Decrypt decrypts a message that was encrypted using the public key
+func (ok *OffchainKeyring) Decrypt(ciphertext []byte) (plaintext []byte, err error) {
+	if len(ciphertext) < box.Overhead {
```
Comment (pszal) [inline core/services/keystore/keys/ocr2key/offchain_keyring.go:55]: needed? I guess, `box.Open()` would validate it.

File: core/services/keystore/keys/ocr2key/offchain_keyring_test.go [hunk 4]
```diff
@@ -16,3 +18,31 @@ func TestOffchainKeyring(t *testing.T) {
...
+
+	decryptedMessage, err := kr.Decrypt(encryptedMessage)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, originalMessage, decryptedMessage)
+}
+
+func Encrypt(peerPublicKey [curve25519.PointSize]byte, plaintext []byte) (ciphertext []byte, err error) {
```
Comment (pszal) [inline core/services/keystore/keys/ocr2key/offchain_keyring_test.go:41]: if that's a test helper pass `t *testing.T` as the first arg, call `t.Helper()` at the beginning of the func, and for every non-nil err call `t.Fatalf()`.

File: core/services/keystore/keys/ocr2key/offchain_keyring.go [hunk 5]
```diff
@@ -47,6 +50,25 @@ func newOffchainKeyring(encryptionMaterial, signingMaterial io.Reader) (*Offchai
 	return ok, nil
 }
 
+// Decrypt decrypts a message that was encrypted using the public key
+func (ok *OffchainKeyring) Decrypt(ciphertext []byte) (plaintext []byte, err error) {
```
Comment (kaleofduty) [inline core/services/keystore/keys/ocr2key/offchain_keyring.go:54]: I would suggest using a more precise name here, e.g. `NaclBoxOpenAnonymous`. After all there are any decryption schemes that could be used here.

File: core/services/keystore/keys/ocr2key/offchain_keyring.go (hunk 2 above)
Comment (KuphJr) [reply core/services/keystore/keys/ocr2key/offchain_keyring.go:64]: I tried both, but decided to go with OpenAnonymous as I couldn't see how the verification would be necessary.  (`nonce` was an artifact I left by mistake).

File: core/services/keystore/keys/ocr2key/key_bundle.go [hunk 6]
```diff
@@ -27,6 +27,7 @@ type KeyBundle interface {
 	Unmarshal(b []byte) (err error)
 	Raw() Raw
 	OnChainPublicKey() string
+	NaclBoxOpenAnonymous(ciphertext []byte) (plaintext []byte, err error)
```
Comment (kaleofduty) [inline core/services/keystore/keys/ocr2key/key_bundle.go:30]: Add a comment explaining which of the many keypairs in the `KeyBundle` is used to open the ciphertext?

File: core/services/keystore/keys/ocr2key/key_bundle.go [hunk 7]
```diff
@@ -26,6 +26,8 @@ type KeyBundle interface {
 	Unmarshal(b []byte) (err error)
 	Raw() Raw
 	OnChainPublicKey() string
+	// Decrypts ciphertext using the encryptionKey from an OCR2 OffchainKeyring
+	NaclBoxOpenAnonymous(ciphertext []byte) (plaintext []byte, err error)
```
Comment (bolekk) [inline core/services/keystore/keys/ocr2key/key_bundle.go:30]: I'm not convinced that this is the best method name, compared to something obvious like "decrypt". Nacl and box are implementation details.

File: core/services/keystore/keys/ocr2key/offchain_keyring.go [hunk 8]
```diff
@@ -47,6 +50,22 @@ func newOffchainKeyring(encryptionMaterial, signingMaterial io.Reader) (*Offchai
 	return ok, nil
 }
 
+// NaclBoxOpenAnonymous decrypts a message that was encrypted using the public key
```
Comment (bolekk) [inline core/services/keystore/keys/ocr2key/offchain_keyring.go:53]: which public key?

File: core/services/keystore/keys/ocr2key/offchain_keyring_test.go [hunk 9]
```diff
@@ -16,3 +18,51 @@ func TestOffchainKeyring(t *testing.T) {
...
+	kr, err := newOffchainKeyring(cryptorand.Reader, cryptorand.Reader)
+	require.NoError(t, err)
+
+	invalid := []byte("invalidEncryptedMessage")
+
+	_, err = kr.NaclBoxOpenAnonymous(invalid)
+	assert.Equal(t, err.Error(), "decryption failed")
+}
+
+func NaclBoxSealAnonymous(t *testing.T, peerPublicKey [curve25519.PointSize]byte, plaintext []byte) []byte {
```
Comment (bolekk) [inline core/services/keystore/keys/ocr2key/offchain_keyring_test.go:58]: lowercase name for a local method

**Review by bolekk — APPROVED** [review]
Looks good to me assuming Research doesn't have any more change requests.

File: core/services/keystore/keys/ocr2key/key_bundle.go (hunk 7 above)
Comment (KuphJr) [reply core/services/keystore/keys/ocr2key/key_bundle.go:30]: This specific name was explicitly requested by Research.

**Pull request merged**